### PR TITLE
Improve redis semantic cache implementation

### LINKED ISSUE #3056: Server failing to start while setting for Redis Semantic Cache [Bug]:
### What happened?

Litellm proxy server is crashing while trying to deploy on the kubernetes cluster.

### Relevant log output

```shell
Getting the following error while deploying the proxy server on kubernetes cluster with semantic cache set as true.

INFO: Started server process [1]
INFO: Waiting for application startup.
ERROR: Traceback (most recent call last):
File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 732, in lifespan
async with self.lifespan_context(app) as maybe_state:
File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 608, in aenter
await self._router.startup()
File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 709, in startup
await handler()
File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 3061, in startup_event
await initialize(**worker_config)
File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 2905, in initialize
) = await proxy_config.load_config(router=llm_router, config_file_path=config)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 2059, in load_config
self._init_cache(cache_params=cache_params)
File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 1960, in _init_cache
litellm.cache = Cache(**cache_params)
^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.11/site-packages/litellm/caching.py", line 1326, in init
self.cache = RedisSemanticCache(
^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.11/site-packages/litellm/caching.py", line 582, in init
from redisvl.index import SearchIndex
ModuleNotFoundError: No module named 'redisvl'

ERROR: Application startup failed. Exiting.

#------------------------------------------------------------#

'This feature doesn't meet my needs because...'
https://github.com/BerriAI/litellm/issues/new
#------------------------------------------------------------#
```

### Twitter / LinkedIn details

_No response_

### LINKED ISSUE #3363: [Bug]:  New schema format introduced; please update schema spec. (type=value_error) [redis-semantic]
### What happened?

Getting stacktrace during app start 

Config:

```
model_list:
  - model_name: gpt-3.5-turbo
    litellm_params:
        model: gpt-3.5-turbo
        api_key: sk-E0BrSPzzi06mfRqy-------9bO
        rpm: 200
  - model_name: gpt-3.5-turbo-16k
    litellm_params:
        model: gpt-3.5-turbo-16k
        api_key: sk-E0BrSPzzi06mfRqy-------9bO
        rpm: 100

general_settings:
  litellm_master_key: sk-aztest
  ui_username: litellm
  ui_password: litellm

litellm_settings:
  set_verbose: True
  cache: True
  cache_params:
    type: "redis-semantic"
    host: "localhost"
    port: "6379"
    password: "password"
    similarity_threshold: 0.8
    redis_semantic_cache_use_async: True
    redis_semantic_cache_embedding_model: "text-embedding-ada-002"
```

### Relevant log output

```shell
❯ litellm --config config.yaml
INFO:     Started server process [42803]
INFO:     Waiting for application startup.

#------------------------------------------------------------#
#                                                            #
#           'It would help me if you could add...'            #
#        https://github.com/BerriAI/litellm/issues/new        #
#                                                            #
#------------------------------------------------------------#

 Thank you for using LiteLLM! - Krrish & Ishaan



Give Feedback / Get Help: https://github.com/BerriAI/litellm/issues/new



Setting Cache on Proxy
redis semantic-cache initializing INDEX - litellm_semantic_cache_index
redis semantic-cache redis_url: redis://:password@localhost:6379
ERROR:    Traceback (most recent call last):
  File "/Users/mda/.pyenv/versions/3.12.2/lib/python3.12/site-packages/starlette/routing.py", line 734, in lifespan
    async with self.lifespan_context(app) as maybe_state:
  File "/Users/mda/.pyenv/versions/3.12.2/lib/python3.12/site-packages/starlette/routing.py", line 610, in __aenter__
    await self._router.startup()
  File "/Users/mda/.pyenv/versions/3.12.2/lib/python3.12/site-packages/starlette/routing.py", line 711, in startup
    await handler()
  File "/Users/mda/.pyenv/versions/3.12.2/lib/python3.12/site-packages/litellm/proxy/proxy_server.py", line 3192, in startup_event
    await initialize(**worker_config)
  File "/Users/mda/.pyenv/versions/3.12.2/lib/python3.12/site-packages/litellm/proxy/proxy_server.py", line 3036, in initialize
    ) = await proxy_config.load_config(router=llm_router, config_file_path=config)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mda/.pyenv/versions/3.12.2/lib/python3.12/site-packages/litellm/proxy/proxy_server.py", line 2060, in load_config
    self._init_cache(cache_params=cache_params)
  File "/Users/mda/.pyenv/versions/3.12.2/lib/python3.12/site-packages/litellm/proxy/proxy_server.py", line 1961, in _init_cache
    litellm.cache = Cache(**cache_params)
                    ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mda/.pyenv/versions/3.12.2/lib/python3.12/site-packages/litellm/caching.py", line 1490, in __init__
    self.cache = RedisSemanticCache(
                 ^^^^^^^^^^^^^^^^^^^
  File "/Users/mda/.pyenv/versions/3.12.2/lib/python3.12/site-packages/litellm/caching.py", line 776, in __init__
    self.index = SearchIndex.from_dict(schema)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mda/.pyenv/versions/3.12.2/lib/python3.12/site-packages/redisvl/index/index.py", line 266, in from_dict
    schema = IndexSchema.from_dict(schema_dict)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mda/.pyenv/versions/3.12.2/lib/python3.12/site-packages/redisvl/schema/schema.py", line 280, in from_dict
    return cls(**data)
           ^^^^^^^^^^^
  File "/Users/mda/.pyenv/versions/3.12.2/lib/python3.12/site-packages/pydantic/v1/main.py", line 341, in __init__
    raise validation_error
pydantic.v1.error_wrappers.ValidationError: 1 validation error for IndexSchema
__root__
  New schema format introduced; please update schema spec. (type=value_error)

ERROR:    Application startup failed. Exiting.
```

### Twitter / LinkedIn details

_No response_

### LINKED ISSUE #4001: [Feature]: Allow Redis Semantic caching with custom Embedding models
### The Feature

Currently, I notice that the schema for the redis semantic cache enforces a dimension size of **1536**. This works well with OpenAI's text-embedding-ada-002 models but fail for any other model. 

```python
schema = {
            "index": {
                "name": "litellm_semantic_cache_index",
                "prefix": "litellm",
                "storage_type": "hash",
            },
            "fields": {
                "text": [{"name": "response"}],
                "text": [{"name": "prompt"}],
                "vector": [
                    {
                        "name": "litellm_embedding",
                        "dims": 1536,
                        "distance_metric": "cosine",
                        "algorithm": "flat",
                        "datatype": "float32",
                    }
                ],
            },
        }
```
Please make the embedding dims configurable from the model-config.yaml file so that a larger range of embedding models deployed with LiteLLM can be used for the cache.

### Motivation, pitch

For caching, much smaller embedding models might be preferred due to their cost and speed. If the dims becomes a configurable item, then this opens up much more interesting modes of caching and cache optimization. It should also reduce costs significantly as this would avoid calls to OpenAI every time a new user query is recieved. 

### Twitter / LinkedIn details

_No response_

### PR DESCRIPTION
**Pull request by tylerhutcherson**
## Relevant issues

- Fixes https://github.com/BerriAI/litellm/issues/3056
- Fixes https://github.com/BerriAI/litellm/issues/3363
- Fixes https://github.com/BerriAI/litellm/issues/4001

## Type
🆕 New Feature
🐛 Bug Fix
🧹 Refactoring
📖 Documentation

## Changes

- Use latest version of RedisVL (`0.3.2`) and fix pydantic & schema version issues
- Support isolated/customized index names to allow for multitenancy
- Support arbitrary embedding model in index schema provided through litellm.embeddings module
- Use built-in `SemanticCache` extension for cleaner code and processing
- Add TTL support to redis semantic cache (OOTB feature of Redis rather than a tack-on)

### CODE REVIEW

File: litellm/caching.py
Comment (rbs333) [inline litellm/caching.py]: Should this todo be merged/public?

**Review by rbs333 — APPROVED** [review]
Generally LGTM had a question about the todo

File: litellm/caching.py
Comment (tylerhutcherson) [reply litellm/caching.py]: Mostly leaving this as a note for the community so they know we will add async support to semantic cache soon as a patch release -- happy to remove if maintainers want it out

Comment (ishaan-jaff) [reply litellm/caching.py]: so redis semantic cache is not async on this PR @tylerhutcherson ?

**Review by ishaan-jaff — CHANGES_REQUESTED** [review]
Is this missing async support @tylerhutcherson ? 

If yes, can you add async support on this PR. Majority of users are trying to use this our llm gateway - which requires all async functions

File: litellm/caching.py
Comment (ishaan-jaff) [inline litellm/caching.py]: so this is a sync call that occurs within `async_get_cache` ?

File: litellm/caching.py
Comment (tylerhutcherson) [reply litellm/caching.py]: The `SemanticCache` class in redisvl library is not async-enabled **yet**. This is on our roadmap for release very soon.

However, this `async_set_cache` method still works and tests pass as expected. Just left a note to update this soon.

Given all of the issues associated with the `redisvl==0.0.7` pin you all have now (non stable version), imo it would be good to get this out and solve all of those open issues. We can prioritize adding in async to our semantic cache class asap. Thoughts @ishaan-jaff ?

Comment (ishaan-jaff) [reply litellm/caching.py]: Async Support is a hard requirement for LiteLLM. Majority of our users use async. If we merge this and users use this in prod their litellm service will go down. We have seen this before when a non-async function was used. 

We're happy to wait on this until redisvl add async support

Comment (ishaan-jaff) [reply litellm/caching.py]: cc @tylerhutcherson

**Comment by tylerhutcherson** [comment]
> Is this missing async support @tylerhutcherson ?
> 
> If yes, can you add async support on this PR. Majority of users are trying to use this our llm gateway - which requires all async functions

Hey @ishaan-jaff @krrishdholakia see my comment here:
https://github.com/BerriAI/litellm/pull/5412#discussion_r1736228236

Let me know what you think!

**Comment by ishaan-jaff** [comment]
@tylerhutcherson 

Async Support is a hard requirement for LiteLLM. Majority of our users use async. If we merge this and users use this in prod their litellm service will go down. We have seen this before when a non-async function was used.

We're happy to wait on this until redisvl add async support

**Comment by tylerhutcherson** [comment]
> @tylerhutcherson
> 
> Async Support is a hard requirement for LiteLLM. Majority of our users use async. If we merge this and users use this in prod their litellm service will go down. We have seen this before when a non-async function was used.
> 
> We're happy to wait on this until redisvl add async support

Understood -- thanks for the clarity. Will see what we can do, thanks!

**Comment by tylerhutcherson** [comment]
FYI https://github.com/redis/redis-vl-python/pull/214 We are close to finalizing support.

**Comment by tylerhutcherson** [comment]
Now that https://github.com/redis/redis-vl-python 0.3.3 is out with async sem cache support.... I have updated this branch to reflect that. Thanks!

@ishaan-jaff @krrishdholakia

**Review by ishaan-jaff — COMMENTED** [review]
looks good - could you send a screenshot of your tests passing for you locally ? Happy to merge after that

**Review by ishaan-jaff — CHANGES_REQUESTED** [review]
looks good - could you send a screenshot of your tests passing for you locally ? Happy to merge after that  @tylerhutcherson

**Comment by tylerhutcherson** [comment]
> looks good - could you send a screenshot of your tests passing for you locally ? Happy to merge after that @tylerhutcherson

Sure thing -- screenshot below. A few caveats:

- There seems to be some issue with async tests here. The poetry env does not include `pytest-asyncio` or anything so the test module skips all async tests and throws warnings like this?

```
litellm/tests/test_caching.py: 20 warnings
  /Users/tyler.hutcherson/Library/Caches/pypoetry/virtualenvs/litellm-g3ANcoZD-py3.11/lib/python3.11/site-packages/_pytest/python.py:183: PytestUnhandledCoroutineWarning: async def functions are not natively supported and have been skipped.
  You need to install a suitable plugin for your async framework, for example:
    - anyio
    - pytest-asyncio
    - pytest-tornasync
    - pytest-trio
    - pytest-twisted
    warnings.warn(PytestUnhandledCoroutineWarning(msg.format(nodeid)))

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
```

- Also I don't have Azure API access nor a dependency called `diskcache` for these tests... so there are 3 tests that fail (seen below). All of the Redis cases pass using `poetry run pytest litellm/tests/test_caching.py`.

Screenshot 2024-09-10 at 2 20 24 PM

@ishaan-jaff

**Pull request merged**
